presto-hive/src/main/java/com/facebook/presto/hive/util/MergingPageIterator.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.util;
+
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.SortOrder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.Queue;
+
+import static com.facebook.presto.hive.util.TempFileWriter.appendPositionTo;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class MergingPageIterator
+        extends AbstractIterator<Page>
+{
+    private final Queue<PagePositionIterator> queue = new PriorityQueue<>();
+    private final List<Type> types;
+    private final List<Integer> sortFields;
+    private final List<SortOrder> sortOrders;
+    private final PageBuilder pageBuilder;
+
+    public MergingPageIterator(
+            Collection<Iterator<Page>> iterators,
+            List<Type> types,
+            List<Integer> sortFields,
+            List<SortOrder> sortOrders)
+    {
+        requireNonNull(sortFields, "sortFields is null");
+        requireNonNull(sortOrders, "sortOrders is null");
+        checkArgument(sortFields.size() == sortOrders.size(), "sortFields and sortOrders size must match");
+
+        this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.sortFields = ImmutableList.copyOf(sortFields);
+        this.sortOrders = ImmutableList.copyOf(sortOrders);
+        this.pageBuilder = new PageBuilder(types);
+
+        iterators.stream()
+                .map(PagePositionIterator::new)
+                .filter(PagePositionIterator::hasNext)
+                .forEach(queue::add);
```

Comment:
it would be better to reuse:
```
https://google.github.io/guava/releases/21.0/api/docs/com/google/common/collect/Iterators.html#mergeSorted-java.lang.Iterable-java.util.Comparator-
```
to construct `PagePosition` iterator.

Reply:
`PagePositionIterator` is not a real `Iterator`. It's a cursor that allows appending the current position to a `PageBuilder`. An alternative approach is make a real iterator of an immutable `PagePosition` class (though that would be a lot of extra object allocation, which may or may not matter). That might simplify the code.